src/coreclr/src/jit/morph.cpp

```diff
@@ -12554,20 +12578,44 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
 
                 if (op2->IsCnsIntOrI() && varTypeIsIntegralOrI(typ))
                 {
-                    /* Fold "((x+icon1)+icon2) to (x+(icon1+icon2))" */
+                    // Fold "((x <op> icon1) <op> icon2) to (x <op> (icon1 <op> icon2))"
                     CLANG_FORMAT_COMMENT_ANCHOR;
 
-                    if (op1->gtOper == GT_ADD &&                             //
-                        !gtIsActiveCSE_Candidate(op1) &&                     //
-                        !op1->gtOverflow() &&                                //
-                        op1->AsOp()->gtOp2->IsCnsIntOrI() &&                 //
-                        (op1->AsOp()->gtOp2->OperGet() == op2->OperGet()) && //
-                        (op1->AsOp()->gtOp2->TypeGet() != TYP_REF) &&        // Don't fold REFs
-                        (op2->TypeGet() != TYP_REF))                         // Don't fold REFs
+                    if (op1->OperIs(oper) &&                                      //
+                        !gtIsActiveCSE_Candidate(op1) &&                          //
+                        op1->AsOp()->gtGetOp2()->IsCnsIntOrI() &&                 //
+                        (op1->AsOp()->gtGetOp2()->OperGet() == op2->OperGet()) && //
+                        (op1->AsOp()->gtGetOp2()->TypeGet() != TYP_REF) &&        // Don't fold REFs
```

Comment:
That is likely should be replaced with `varTypeIsGC` check so it would block `byref` as well.